File: tests/testthat/test-bmi_code.R
library(testthat)

# Test for calculate_bmi
test_that("calculate_bmi works correctly", {
  expect_equal(calculate_bmi(1.80, 75), 75 / (1.80^2))
  expect_error(calculate_bmi(0, 75), "Height must be greater than zero.")
  expect_error(calculate_bmi(1.80, 0), "Weight must be greater than zero.")
  expect_equal(calculate_bmi(1.70, 65), 10)
})

# Test for calculate_bmi_df
test_that("calculate_bmi_df works correctly", {
  df <- data.frame(height = c(1.80, 1.60), weight = c(75, 60))
  result <- calculate_bmi_df(df)
  expected_bmi <- c(75 / (1.80^2), 60 / (1.60^2))
  expect_equal(result$bmi, expected_bmi)

  df_missing <- data.frame(h = c(1.80), w = c(75))
  expect_error(calculate_bmi_df(df_missing), "Dataframe must contain 'height' and 'weight' columns.")
})

File: tests/python_best_practices/test_bmi_code.py
import pytest
import pandas as pd
from python_best_practices.bmi import calculate_bmi, calculate_bmi_df


def test_calculate_bmi_valid():
    assert calculate_bmi(1.80, 75) == pytest.approx(75 / (1.80**2))


def test_calculate_bmi_invalid_height():
    with pytest.raises(ValueError, match="Height must be greater than zero."):
        calculate_bmi(0, 75)


def test_calculate_bmi_invalid_weight():
    with pytest.raises(ValueError, match="Weight must be greater than zero."):
        calculate_bmi(1.80, 0)


def test_calculate_bmi_df_valid():
    df = pd.DataFrame({"height": [1.80, 1.60], "weight": [75, 60]})
    result = calculate_bmi_df(df)
    expected_bmi = [75 / (1.80**2), 60 / (1.60**2)]
    assert result["bmi"].tolist() == pytest.approx(expected_bmi)


def test_calculate_bmi_df_missing_columns():
    df = pd.DataFrame({"h": [1.80], "w": [75]})
    with pytest.raises(
        ValueError,
        match="DataFrame must contain 'height' and 'weight' columns.",
    ):
        calculate_bmi_df(df)
